src/Microsoft.OData.Core/UriParser/ODataPathInfo.cs

```diff
@@ -44,6 +46,19 @@ public ODataPathInfo(ODataPath odataPath)
                         this.targetEdmType = collectionType.ElementType.Definition;
                     }
                 }
+
+                // If the last segment is an OperationSegment for a bound operation and the targetNavigationSource is null,
+                // We use the targetNavigationSource for the previous segment.
+                // e.g People/My.Function.GetPeopleWithDogs()
+                if (this.targetNavigationSource == null && lastSegment is OperationSegment operationSegment)
+                {
+                    IEdmOperation operation = operationSegment.Operations.FirstOrDefault();
+
+                    if (operation.IsBound & previous != null)
+                    {
+                        this.targetNavigationSource = previous.TargetEdmNavigationSource;

```

Comment:
Could `previous.TargetEdmNavigationSource` also be null, or that isn't an issue?

Reply:
What if the return type of the of the operation does not match the navigation source of the previous segment?
For example, what if we have something like `Customers/NS.GetTopOrders()?$expand=Products($filter=Category eq 'Electronics')`  where `GetTopOrders` returns a collection of `Order` entities from an `Orders` entity set, which is different from `Customers` entity set?

Reply:
I don't think we should delegate the targetNS to the NS of preview segment.
I'd figure out why does ODL needs the targetNS? If the last operation doesn't have the targetNS, can we use the return type of the operation to do something? I do think we use the TargetNS to calculate the context URL, if we don't have the NS, we should use the type to generate the Context URL. @mikepizzo